mycpp/gc_dict.h

```diff
@@ -307,49 +275,119 @@ void Dict<K, V>::clear() {
   len_ = 0;
 }
 
-// Returns the position in the array.  Used by dict_contains(), index(),
-// get(), and set().
-//
-// For now this does a linear search.
 // TODO:
-// - hash functions, and linear probing.
-// - resizing based on load factor
-//   - which requires rehashing (re-insert all items)
 // - Special case to intern Str* when it's hashed?  How?
 //   - Should we have wrappers like:
 //   - V GetAndIntern<V>(D, &string_key)
 //   - SetAndIntern<V>(D, &string_key, value)
 //   This will enable duplicate copies of the string to be garbage collected
 template <typename K, typename V>
-int Dict<K, V>::position_of_key(K key) {
-  for (int i = 0; i < capacity_; ++i) {
-    int special = entry_->items_[i];  // NOT an index now
-    if (special == kDeletedEntry) {
-      continue;  // keep searching
+int Dict<K, V>::find_key_in_index(K key) const {
+  if (capacity_ == 0) {
+    return -1;
+  }
+
+  // Hash the key onto a slot in the index. If the first slot is occupied, probe
+  // until an empty one is found.
+  unsigned h = hash_key(key);
+  int init_bucket = h % capacity_;
+
+  // If there's a vacancy because of deletion on the probe path for this key, we
+  // should fill it before consuming a slot that has never been used. This
+  // should help keep the index somewhat compact.
+  int tombstone = -1;
+  for (int i = init_bucket; i < capacity_; ++i) {
+    int pos = entry_->items_[i];  // NOT an index now
+    DCHECK(pos < len_);
+    if (pos == kDeletedEntry) {
+      if (tombstone == -1) {
+        tombstone = pos;
+      }
+      continue;
     }
-    if (special == kEmptyEntry) {
-      return -1;  // not found
+    if (pos == kEmptyEntry) {
+      if (tombstone != -1) {
+        return tombstone;
+      }
+      return i;
     }
+    unsigned h2 = hash_key(keys_->items_[pos]);
+    if (h == h2 && keys_equal(keys_->items_[pos], key)) {
+      return i;
+    }
+  }
+
+  // Didn't find anything. Try wrapping around.
+  for (int i = 0; i < init_bucket; ++i) {
+    int pos = entry_->items_[i];  // NOT an index now
+    DCHECK(pos < len_);
+    if (pos == kDeletedEntry) {
+      if (tombstone == -1) {
+        tombstone = pos;
+      }
+      continue;
+    }
+    if (pos == kEmptyEntry) {
+      if (tombstone != -1) {
+        return tombstone;
+      }
+      return i;
+    }
+    unsigned h2 = hash_key(keys_->items_[pos]);
+    if (h == h2 && keys_equal(keys_->items_[pos], key)) {
+      return i;
+    }
+  }
+
+  return tombstone;
+}
+
+template <typename K, typename V>
+int Dict<K, V>::find_key(K key) const {
+  if (entry_ != nullptr) {
+    // Common case.
+    int pos = find_key_in_index(key);
+    if (pos == -1 || entry_->items_[pos] < 0) {
+      return -1;
+    }
+    return entry_->items_[pos];
+  }
+
+  // GlobalDict. Just scan.
+  for (int i = 0; i < len_; ++i) {
     if (keys_equal(keys_->items_[i], key)) {
       return i;
     }
   }
-  return -1;  // table is completely full?  Does this happen?
+
+  // Not found.
```

Comment:
do we need a kNotFound ?

maybe it can be distinct from kDeletedEntry and kEmptyEntry, so it's easier to see the difference in a debugger

It could be -3 I think